rust-runtime/aws-smithy-http/src/header.rs

```diff
@@ -164,43 +164,106 @@ where
     }
 }
 
+/// Functions for parsing multiple comma-delimited header values out of a
+/// single header. This parsing adheres to
+/// [RFC-7230's specification of header values](https://datatracker.ietf.org/doc/html/rfc7230#section-3.2.6).
+mod parse_multi_header {
+    use super::ParseError;
+    use std::borrow::Cow;
+
+    /// Reads a single value out of the given input, and returns a tuple containing
+    /// the parsed value and the remainder of the slice that can be used to parse
+    /// more values.
+    pub(crate) fn read_value(input: &[u8]) -> Result<(Cow<'_, str>, &[u8]), ParseError> {
+        for (index, &byte) in input.iter().enumerate() {
+            let current_slice = &input[index..];
+            match byte {
+                b' ' | b'\t' => { /* skip whitespace */ }
+                b'"' => return read_quoted_value(&current_slice[1..]),
+                _ => return read_unquoted_value(current_slice),
+            }
+        }
+
+        // We only end up here if the entire header value was whitespace or empty
+        Ok((Cow::Borrowed(""), &[]))
+    }
+
+    fn read_unquoted_value(input: &[u8]) -> Result<(Cow<'_, str>, &[u8]), ParseError> {
+        let next_delim = input.iter().position(|&b| b == b',').unwrap_or(input.len());
+        let (first, next) = input.split_at(next_delim);
+        let first = std::str::from_utf8(first)
+            .map_err(|_| ParseError::new_with_message("header was not valid utf8"))?;
+        Ok((Cow::Borrowed(first), then_delim(next).unwrap()))
+    }
+
+    /// Reads a header value that is surrounded by quotation marks and may have escaped
+    /// quotes inside of it.
+    fn read_quoted_value(input: &[u8]) -> Result<(Cow<'_, str>, &[u8]), ParseError> {
+        for index in 0..input.len() {
+            match input[index] {
+                b'"' if index == 0 || input[index - 1] != b'\\' => {
+                    let mut inner =
+                        Cow::Borrowed(std::str::from_utf8(&input[0..index]).map_err(|_| {
+                            ParseError::new_with_message("header was not valid utf8")
+                        })?);
+                    if inner.contains("\\\"") {
+                        inner = Cow::Owned(inner.replace("\\\"", "\""));
+                    }
+                    let rest = then_delim(&input[(index + 1)..])?;
+                    return Ok((inner, rest));
+                }
+                _ => {}
+            }
+        }
+        Err(ParseError::new_with_message(
+            "header value had quoted value without end quote",
+        ))
+    }
+
+    fn then_delim(s: &[u8]) -> Result<&[u8], ParseError> {
```

Comment:
`then_comma` to clarify that we're looking for commas and not quotes (since they're both delimiters)?